cimi-resources/src/com/sixsq/slipstream/ssclj/resources/external_object/utils.clj

```diff
@@ -45,7 +51,109 @@
       content-type (.setContentType req content-type))
     (str (.generatePresignedUrl (get-s3-client obj-store-conf) req))))
 
+
 (defn delete-s3-object [obj-store-conf bucket obj-name]
   (let [deleteRequest (DeleteObjectRequest. bucket obj-name)]
     (.deleteObject (get-s3-client obj-store-conf) deleteRequest)))
 
+
+(defn bucket-exists?
+  "Function mocked in unit tests"
+  [s3client bucketName]
+  (.doesBucketExist s3client bucketName))
+
+
+(defn create-bucket!
+  "Caller should have checked that bucket does not exist yet. If creation
+  fails, an Exception is thrown; otherwise true is returned. Mocked in unit
+  tests."
+  [s3client bucket-name]
+  (.createBucket s3client (CreateBucketRequest. bucket-name))
+  true)
+
+
+(defn expand-cred
+  "Returns credential document after expanding `href-obj-store-cred` credential href.
+
+  Deriving objectType from request is not directly possible as this is a request on
+  action resource. We would need to get resource id, load the resource and get
+  objectType from it. Instead, requiring objectType as parameter. It should be known
+  to the callers."
+  [href-obj-store-cred]
+  (std-crud/resolve-hrefs href-obj-store-cred request-admin true))
+
+;;Pre-conditions for adding external-object
+
+
+(defn bucket-creation-ok?
+  "When the requested bucket doesn't exist and can't be created.
+  The external object resource must not be created."
+  [s3client bucketName]
+  (try
+    (or (bucket-exists? s3client bucketName)
+        (create-bucket! s3client bucketName))
+    (catch Exception e
+      (log/error (format "Error when creating bucket %s: %s" bucketName (.getMessage e)))
+      false)))
+
+
+(defn authorized-bucket-operation?
+  [{:keys [objectStoreCred] :as resource} request check-authn-fn]
+  (-> objectStoreCred
+      (expand-cred)
+      (check-authn-fn request)))
+
+
+(defn uploadable-bucket?
+  "When the bucket exists, but the user can't create the object.
+  The external object resource must not be created."
+  [obj-store-conf bucketName]
+  (let [upload-url (generate-url obj-store-conf bucketName "test-probe" :put {:ttl 3 :content-type "text/plain" :filename "test.txt"})]
+    (= 200 (http/put upload-url))))
+
```

Comment:
I don't think that this implementation will work in practice.  The test file is never deleted. The second time the user tries to add an object to the same bucket, this request will fail and cause the function to incorrectly return false.  A cleaner solution with fewer error corner cases would be to check the ACL of the bucket itself, provided that the same calls work at both Amazon and Exoscale.

Reply:
https://docs.aws.amazon.com/AmazonS3/latest/dev/acl-using-java-sdk.html

Reply:
Implementation changed in favor a of a HEAD request to the bucket. As stated by the Javadoc (https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/AmazonS3Client.html#headBucket-com.amazonaws.services.s3.model.HeadBucketRequest-) "This operation is useful to determine if a bucket exists and you have permission to access it."